---
title: 'BCB420 A2: Differential Gene Expression and Preliminary ORA'
author: "Kara Han"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_depth: '2'
    df_print: paged
---

```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
```

# Introduction: A1 Results Summary

For Assignment 1, I selected the GEO dataset with accession number GSE166605: 
RNA-sequencing analysis of effect of miR-155 overexpression in human CD14+ monocytes
(Olsson et al. 2021).

MicroRNAs (miRs) are essential regulators for pro-inflammatory effector functions 
of myeloid cells and miR dysfunction is implicated in rheumatoid arthritis (RA).
The study investigates two groups, miR-155 and negative control, in 5 donors to 
understand the effect of miR-155 overexpression in human CD14+ monocytes. 

The raw count data contains a total of 60603 genes. Genes that have low counts 
were removed from the raw count data and normalization was subsequently performed 
using the `edgeR` package. The multidimensional scaling (MDS) plot of the 
post-normalization data showed a clear separation of the samples in the two 
groups, suggesting that miR-155 groups and negaitve control groups have
different behaviours across their counts and differential expressions. The Ensembl
gene IDs in the original count data were mapped to HGNC symbols to create the final
normalized dataset. The final coverage of the miR-155 dataset is 13003 genes along
with their Ensembl gene IDs, HGNC symbols, and normalized counts in 5 donors 
(10 samples).

# Differential Gene Expression Analysis

First, I'll load the required packages and read in the original and normalized 
count datasets.

```{r load_packages, eval=FALSE}
# The code in this report is adapted from Dr. Ruth Isserlin's Lecture 6-7 slides (Isserlin, 2022).

# Install required packages
if (!require("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}

if (!requireNamespace("biomaRt", quietly = TRUE)) {
  BiocManager::install("biomaRt")
}

if (!requireNamespace("GEOquery", quietly = TRUE)) {
  BiocManager::install("GEOquery")
}

if (!requireNamespace("edgeR", quietly = TRUE)) {
  BiocManager::install("edgeR")
}

if (!requireNamespace("limma", quietly = TRUE)) {
  BiocManager::install("limma")
}

if (!requireNamespace("Biobase", quietly = TRUE)) {
  BiocManager::install("Biobase")
}

if (!require("dplyr", quietly = TRUE)) {
  install.packages("dplyr")
}

if (!require("circlize", quietly = TRUE)) {
  install.packages("circlize")
}

if (!require("gprofiler2", quietly = TRUE)) {
  install.packages("gprofiler2")
}

if (!require("plotfunctions", quietly = TRUE)) {
  install.packages("plotfunctions")
}

```

```{r load_count_data}
# A way to download the dataset only when necessary instead of re-downloading it every time the Notebook is run
gseDirectory <- file.path(getwd(), "GSE166605")
if (!file.exists(gseDirectory)) {
  # Get the supplementary files from GEO
  suppFiles <- GEOquery::getGEOSuppFiles("GSE166605")
  # There is only one supplementary file
  fnames <- rownames(suppFiles)
} else {
  fnames <- file.path(gseDirectory, "GSE166605_D1-5_Neg_C_miR_155_RAW_counts.txt.gz")
}

# Read in the original count data
expData <- read.delim(fnames[1], header = TRUE, check.names = FALSE)
head(expData)

# Read in the normalized count data
normalizedCountsData <- read.table(file = file.path(getwd(), "GSE166605_normalized_counts.txt"),
                                   header = TRUE, sep = "\t",
                                   stringsAsFactors = FALSE,
                                   check.names = FALSE)
head(normalizedCountsData)
```

## Data Visualization

Let's use the multidimensional scaling (MDS) plot to decide which factors to
include in our model.

```{r}
# Translate raw counts into counts per million (CPM)
cpms <- edgeR::cpm(expData[, 3:12])
rownames(cpms) <- expData[, 1]

# Keep the genes with high enough counts
keep <- rowSums(cpms > 1) >= 5
expDataFiltered <- expData[keep, ] # 13003 observations

filteredDataMatrix <- as.matrix(expDataFiltered[, 3:12])
rownames(filteredDataMatrix) <- expDataFiltered$GeneID

# Define the cell types
# Extract the donor and cell types from the original data
samples <- data.frame(lapply(colnames(expData)[3:12], 
                             FUN = function(x) {unlist(strsplit(x, split = " "))[c(1, 3)]}))
colnames(samples) <- colnames(expData)[3:12]
rownames(samples) <- c("donors", "cell_type")
samples <- data.frame(t(samples))

# Subset the normalized data to get expression values only
expressionMatrix <- as.matrix(normalizedCountsData[, c(3:ncol(normalizedCountsData))])

# MDS plot of samples, colored by cell type
mdsCellType <- limma::plotMDS(expressionMatrix,
                              pch = 1,
                              col = c("darkgreen", "blue")[factor(samples$cell_type)],
                              main = "MDS Plot Colored by Cell Type")

legend("bottomleft",
       legend = levels(factor(samples$cell_type)),
       pch = c(1),
       col = c("darkgreen", "blue"),
       title = "Cell Type",
       cex = 0.75)

# MDS plot of samples, colored by donor
mdsDonor <- limma::plotMDS(expressionMatrix,
                           pch = 1,
                           col = c("darkgreen", "blue", "red", "orange", "black")[factor(samples$donors)],
                           main = "MDS Plot Colored by Donor")

legend("bottomleft",
       legend = levels(factor(samples$donors)),
       pch = c(1),
       col = c("darkgreen", "blue", "red", "orange", "black"),
       title = "Donor",
       cex = 0.75)

```


From the MDS plot colored by cell type, we see that there is a clear separation 
between the cell types, suggesting the dissimilarity between miR-155 positive and 
negative groups.

The MDS plot colored by donor suggests that the difference in differential 
expression of genes among different donors is not too clear. But the initial
heatmap of the data seems to suggest that donor differences do play a role in
gene expression. 

As such, we'll develop two models: one only with cell type as the factor and the 
other with both cell type and donor as the factors with the `limma` package.



## P-Value & Multiple Hypothesis Correction

In the code below, we create two linear models: one with cell type as the only
factor, and the other with donor variability added. 

```{r multiple_hypothesis_correction}
# Create a linear model with data defined by cell type only
simpleModelDesign <- model.matrix(~ samples$cell_type)
head(simpleModelDesign)

# Use the heatmapMatrix from earlier to create the data matrix
minimalSet <- Biobase::ExpressionSet(assayData = as.matrix(expressionMatrix))

# Fit the data
fit <- limma::lmFit(minimalSet, simpleModelDesign)

# Use empirical Bayes to compute differential expression
# Use trend = TRUE for RNA-seq data
fit2 <- limma::eBayes(fit, trend = TRUE)

# Find the top fit
# Use Benjamini-Hochberg for multiple hypothesis correction
topfit <- limma::topTable(fit2,
                          coef = ncol(simpleModelDesign),
                          adjust.method = "BH", 
                          number = nrow(expressionMatrix))

# Merge the top fit data with HGNC symbols
simpleModelOutput <- merge(normalizedCountsData[, c(1:2)],
                           topfit,
                           by.x = 1,
                           by.y = 0,
                           all.y = TRUE)

# Sort the output by p-values 
simpleModelOutput <- simpleModelOutput[order(simpleModelOutput$P.Value), ]
head(simpleModelOutput)

# Number of genes that pass the threshold p-value < 0.05
length(which(simpleModelOutput$P.Value < 0.05))

# Number of genes that pass correction
length(which(simpleModelOutput$adj.P.Val < 0.05))

# A more complex model that accounts for donor variability
donorModelDesign <- model.matrix(~ samples$donors + samples$cell_type)
head(donorModelDesign)

# Fit the data
fitDonor <- limma::lmFit(minimalSet, donorModelDesign)

# Use empirical Bayes to compute differential expression
# Use trend = TRUE for RNA-seq data
fit2Donor <- limma::eBayes(fitDonor, trend = TRUE)

# Find the top fit
# Use Benjamini-Hochberg for multiple hypothesis correction
topfitDonor <- limma::topTable(fit2Donor,
                               coef = ncol(donorModelDesign),
                               adjust.method = "BH", 
                               number = nrow(expressionMatrix))

# Merge the top fit data with HGNC symbols
donorModelOutput <- merge(normalizedCountsData[, c(1:2)],
                          topfitDonor,
                          by.x = 1,
                          by.y = 0,
                          all.y = TRUE)
# Order the result by p-values
donorModelOutput <- donorModelOutput[order(donorModelOutput$P.Value), ]
head(donorModelOutput)

# Number of genes that pass the threshold p-value < 0.05
length(which(donorModelOutput$P.Value < 0.05))

# Number of genes that pass correction
length(which(donorModelOutput$adj.P.Val < 0.05))
```

We'll use a volcano plot to demonstrate the amount of differentially expressed
genes.

```{r}
# Plot all the differential expression values
plot(donorModelOutput$logFC, -log10(donorModelOutput$P.Value), pch = 20, 
     main = "Volcano Plot of miR-155 Differential Expression", xlim = c(-10, 10),
     xlab = "logFC", ylab = "-log10(P-value)")

# Subset the upregulated genes and colour them as red
upregulated <- donorModelOutput[which(donorModelOutput$P.Value < 0.05 & donorModelOutput$logFC > 0), ]
with(upregulated, points(logFC, -log10(P.Value), pch = 20, col = "red"))

# Subset the downregulated genes and colour them as blue
downregulated <- donorModelOutput[which(donorModelOutput$P.Value < 0.05 & donorModelOutput$logFC < 0), ]
with(downregulated, points(logFC, -log10(P.Value), pch = 20, col = "blue"))

legend("topright", 
       legend = c("up-regulated", "down-regulated", "neither"), 
       fill = c("red","blue", "black"),
       cex = 1)
```

```{r top_hits_heatmap}
# Find the significantly expressed genes
topHits <- donorModelOutput$ensembl_gene_id[donorModelOutput$P.Value < 0.05]

# Highlight the top hits
heatmapMatrixTopHits <- scale(expressionMatrix[which(rownames(expressionMatrix) %in% topHits), ])

# Create the heatmap for the top hits
topHitsHeatmap <- heatmap(as.matrix(heatmapMatrixTopHits), 
                          Rowv = NA, 
                          Colv = NA,
                          col = cm.colors(256), 
                          margins = c(12, 8),
                          main = "Heatmap of Top Hits using Limma",
                          xlab = "Samples",
                          ylab = "Genes",
                          scale = "row")
# Add a legend
plotfunctions::gradientLegend(valRange = c(min(heatmapMatrixTopHits), max(heatmapMatrixTopHits)),
               color = "cm",
               pos = 0.4, 
               side = 2,
               dec = 0)
```

```{r}
# Grouping the samples for each experimental condition (miR-155 or negative control) together
heatmapTopHitsPosNeg <- heatmapMatrixTopHits[, c(grep(colnames(heatmapMatrixTopHits), pattern = "*miR-155"),
                                                 grep(colnames(heatmapMatrixTopHits), pattern = "*Neg"))]

# Plot the heatmap
topHitsPosNegHeatmap <- heatmap(as.matrix(heatmapTopHitsPosNeg), 
                                Rowv = NA, 
                                Colv = NA,
                                col = cm.colors(256), 
                                margins = c(9, 5),
                                main = "Heatmap of top hits using Limma",
                                xlab = "Samples",
                                ylab = "Genes",
                                scale = "row")

# Add a legend
plotfunctions::gradientLegend(valRange = c(min(heatmapTopHitsPosNeg), max(heatmapTopHitsPosNeg)),
               color = "cm",
               pos = 0.4, 
               side = 2,
               dec = 0)
```

## Summary
* Calculate p-values for each of the genes in your expression set. How many genes were significantly differentially expressed? What thresholds did you use and why?

  + In the simple model with only cell type as the factor, 3280 genes were significantly
  differentially expressed while 5953 genes were significantly differentially expressed
  in the more complex model with donor variability added as another factor. I used
  0.05 as the threshold because 1) the heatmap using 0.01 as the threshold is pretty
  similar to the one with 0.05 as the threshold, thus setting the threshold to 0.05
  would include more genes with weak signals that may contribute to the whole; 2)
  0.05 is used as the threshold in the original publication.

* Multiple hypothesis testing - correct your p-values using a multiple hypothesis correction method. Which method did you use? And Why? How many genes passed correction?

  + I used the Benjamini-Hochberg correction method because it is more commonly
  adapted in multiple hypothesis testing and is a powerful tool used to decrease
  false discovery rate. In the simple model with only cell type as the factor,
  734 genes passed correction. In the more complex model with donor variability,
  more genes passed correction with a total of 3523.

* Show the amount of differentially expressed genes using an MA Plot or a Volcano plot. Highlight genes of interest.
  + See "Volcano Plot of miR-155 Differential Expression" above.

* Visualize your top hits using a heatmap. Do you conditions cluster together? Explain why or why not.
  + The conditions don't seem to cluster together by observing the two types of
  heatmap presented above. Instead, the heatmap suggests a donor-centric pattern:
  most genes are up-regulated in donor 1 while most genes are down-regulated in
  donor 4. This observation is more obvious when we transition from the 
  donor-arranged heatmap to the condition-arranged heatmap. The general trend,
  however, shows that genes are more down-regulated in the miR-155 present condition
  while they are more enriched in the negative controls.

# Thresholded Over-Representation Analysis

Let's first create the thresholded up-regulated and down-regulated gene lists
to be used in the over-representation analysis. 

```{r thresholded_gene_lists}
# Number of genes that are up-regulated
length(which(donorModelOutput$P.Value < 0.05 & donorModelOutput$logFC > 0))

# Number of genes that are down-regulated
length(which(donorModelOutput$P.Value < 0.05 & donorModelOutput$logFC < 0))

# The resulting rank value would be higher if the p-value is smaller
donorModelOutput[, "rank"] <- -log(donorModelOutput$P.Value, base = 10) * sign(donorModelOutput$logFC)
donorModelOutput <- donorModelOutput[order(donorModelOutput$rank), ]

# Create up-regulated gene list
upregulatedGenes <- donorModelOutput[which(donorModelOutput$P.Value < 0.05 & donorModelOutput$logFC > 0), ]

# Create down-regulated gene list
downregulatedGenes <- donorModelOutput[which(donorModelOutput$P.Value < 0.05 & donorModelOutput$logFC < 0), ]
```

Now that we have these two lists of up-regulated and down-regulated genes, we
want to know whether there are any gene sets or pathways that are over-represented
or under-represented in these lists. I will use `g:Profiler` to conduct the
over-representation analysis.

```{r}
# Obtain the version of the annotation data
versionInfo <- gprofiler2::get_version_info(organism = "hsapiens")

versionDF <- as.data.frame(unlist(versionInfo$sources))
names(versionDF)[1] <- "version"
versionDF[, "annotation_data"] <- row.names(versionDF)
versionDF <- versionDF[versionDF$annotation_data %in% c("GO:BP.version", "KEGG.version", "REAC.version", "WP.version"), ]

# Reformat the annotation data version table for a clearer presentation
versionDF <- versionDF[, c(2, 1)]
versionDF$annotation_data <- c("GO:BP", "KEGG", "REAC", "WP")
row.names(versionDF) <- 1:4
head(versionDF)

# Conduct ORA
gostUp <- gprofiler2::gost(query = upregulatedGenes$ensembl_gene_id,
                           significant = FALSE,
                           user_threshold = 0.05,
                           correction_method = "fdr",
                           sources = c("GO:BP", "KEGG", "REAC", "WP"))
nrow(gostUp$result)

gostDown <- gprofiler2::gost(query = downregulatedGenes$ensembl_gene_id,
                             significant = FALSE,
                             user_threshold = 0.05,
                             correction_method = "fdr",
                             sources = c("GO:BP", "KEGG", "REAC", "WP"))

nrow(gostDown$result)

gostAll <- gprofiler2::gost(query = donorModelOutput$ensembl_gene_id,
                             significant = FALSE,
                             user_threshold = 0.05,
                             correction_method = "fdr",
                             sources = c("GO:BP", "KEGG", "REAC", "WP"))
nrow(gostAll$result)
```

There are >10,000 gene sets returned in each ORA, these include root terms that
include many branches of more specific terms. We'll control the size of the gene
set to be smaller.

```{r}
gostUpResults <- gostUp$result[which(gostUp$result$term_size <= 200), ]
nrow(gostUpResults)

gostDownResults <- gostDown$result[which(gostDown$result$term_size <= 200), ]
nrow(gostDownResults)

gostAllResults <- gostAll$result[which(gostAll$result$term_size <= 200), ]
nrow(gostAllResults)
```

Let's visualize the ORA results.

```{r, fig.cap="ORA Results Visualization for All Differentially Expressed Genes"}
gprofiler2::gostplot(gostAll, interactive = FALSE)
```




```{r, fig.cap="ORA Results Visualization for Upregulated Genes"}
gprofiler2::gostplot(gostUp, interactive = FALSE)
```





```{r, fig.cap="ORA Results Visualization for Downregulated Genes"}
gprofiler2::gostplot(gostDown, interactive = FALSE)
```



* Which method did you choose and why?
  + I chose to use the `gprofiler2` R package to conduct my over-representation
  analysis. I found `g:Profiler` quite easy to use from the homework assignment 
  and I really like its design and output visualization. In addition, `g:Profiler`
  is frequently updated to verify that all the data works so it ensures that the
  information we are using is up-to-date.
  
* What annotation data did you use and why? What version of the annotation are you using?
  + I used GO: Biological Pathways, KEGG, Reactome, and WikiPathways as my annotation
  data because I'm interested in analyzing the enrichment in different biological
  pathways. Please see above for the version information.

* How many genesets were returned with what thresholds?
  + With 0.05 as the threshold, 12686 gene sets were returned for the up-regulated
  genes; 13058 gene sets were returned for the down-regulated genes; and 17520
  gene sets for all differentially expressed genes in the dataset.

* Run the analysis using the up-regulated set of genes, and the down-regulated set of genes separately. How do these results compare to using the whole list (i.e all differentially expressed genes together vs. the up-regulated and down regulated differentially expressed genes separately)?

```{r}
# Obtain the top results from each annotation data source for:
# up-regulated genes
upTopResults <- gostUpResults[which(gostUpResults$source == "GO:BP"), ][1, ]
upTopResults <- rbind(upTopResults, gostUpResults[which(gostUpResults$source == "KEGG"), ][1, ])
upTopResults <- rbind(upTopResults, gostUpResults[which(gostUpResults$source == "REAC"), ][1, ])
upTopResults <- rbind(upTopResults, gostUpResults[which(gostUpResults$source == "WP"), ][1, ])
upTopResults <- subset(upTopResults, select = c(9:11))
row.names(upTopResults) <- c(1:4)
upTopResults
```

The up-regulated genes appear to be linked to the ribosome-mediated cytoplasmic
translation of proteins. Ribosome contains rRNAs, and the top REAC result "rRNA
processing" further indicates that the up-regulated genes might be involved in 
the processes that ensure proper ribosome formation in the cytoplasm. 

```{r}
# down-regulated genes
downTopResults <- gostDownResults[which(gostDownResults$source == "GO:BP"), ][1, ]
downTopResults <- rbind(downTopResults, gostDownResults[which(gostDownResults$source == "KEGG"), ][1, ])
downTopResults <- rbind(downTopResults, gostDownResults[which(gostDownResults$source == "REAC"), ][1, ])
downTopResults <- rbind(downTopResults, gostDownResults[which(gostDownResults$source == "WP"), ][1, ])
downTopResults <- subset(downTopResults, select = c(9:11))
row.names(downTopResults) <- c(1:4)
downTopResults
```

Based on the results, the down-regulated genes are involved in pattern recognition
receptor singaling pathway, which includes interferon signaling that is involved 
in protecting the body against viral infections. 

```{r}
allTopResults <- gostAllResults[which(gostAllResults$source == "GO:BP"), ][1, ]
allTopResults <- rbind(allTopResults, gostAllResults[which(gostAllResults$source == "KEGG"), ][1, ])
allTopResults <- rbind(allTopResults, gostAllResults[which(gostAllResults$source == "REAC"), ][1, ])
allTopResults <- rbind(allTopResults, gostAllResults[which(gostAllResults$source == "WP"), ][1, ])
allTopResults <- subset(allTopResults, select = c(9:11))
row.names(allTopResults) <- c(1:4)
allTopResults
```

The results for all differentially expressed genes are more general compared to
the up-regulated and down-regulated gene sets. These genes are generally involved 
in RNA processing and ubiqutin mediated proteolysis.

# Interpretation
1. Do the over-representation results support conclusions or mechanism discussed
in the original paper?

In the original paper, the authors found that over-representation of miR-155 is involved in the expression of pro-inflammatory cytokines, antigen-presenting and co-stimulatory molecules, and negative feedback molecules that limit inflammatory function. Our over-representation results are not the same, but the up-regulated genes could be involved in the formation of the antigen-presenting, co-stimulatory, and negative feedback molecules since they are involved in ribosome-mediate pathways.


2. Can you find evidence, i.e. publications, to support some of the results that
you see. How does this evidence support your results?

I couldn't find much evidence to support my results. miR-155, however, was found to be involved in the degree of inflammatory activation and drives the production of inflammatory cytokines and integrin expression, preventing the switch to anti-inflammatory phenotypes and thus closely related to autoimmune diseases including Rheumatoid arthritis (RA) (Alivernini et al., 2018). I think I might not have chosen the best method for the differential expression or over-representation analysis. 

# References
1. Olsson, A. M., Povoleri, G., Somma, D., Ridley, M. L., Rizou, T., Lalnunhlimi, S., Macdonald, L., Rajasekhar, M., Martinez-Nunez, R. T., Kurowska-Stolarska, M., & Taams, L. S. (2021). miR-155 overexpressing monocytes resemble HLA highISG15 + synovial tissue macrophages from patients with rheumatoid arthritis and induce polyfunctional CD4+ T cell activation. Clinical and experimental immunology, uxab016. Advance online publication. https://doi.org/10.1093/cei/uxab016

2. Martin Morgan (2021). BiocManager: Access the Bioconductor Project Package Repository. R package version 1.30.16.
https://CRAN.R-project.org/package=BiocManager.

3. Mapping identifiers for the integration of genomic datasets with the R/Bioconductor package biomaRt. Steffen
  Durinck, Paul T. Spellman, Ewan Birney and Wolfgang Huber, Nature Protocols 4, 1184-1191 (2009).

4. BioMart and Bioconductor: a powerful link between biological databases and microarray data analysis. Steffen
Durinck, Yves Moreau, Arek Kasprzyk, Sean Davis, Bart De Moor, Alvis Brazma and Wolfgang Huber, Bioinformatics 21,
3439-3440 (2005).

5. Davis, S. and Meltzer, P. S. GEOquery: a bridge between the Gene Expression Omnibus (GEO) and BioConductor.
Bioinformatics, 2007, 14, 1846-1847.

6. Robinson MD, McCarthy DJ and Smyth GK (2010). edgeR: a Bioconductor package for differential expression analysis
of digital gene expression data. Bioinformatics 26, 139-140.

7. McCarthy DJ, Chen Y and Smyth GK (2012). Differential expression analysis of multifactor RNA-Seq experiments with
respect to biological variation. Nucleic Acids Research 40, 4288-4297

8. Chen Y, Lun ATL, Smyth GK (2016). From reads to genes to pathways: differential expression analysis of RNA-Seq
experiments using Rsubread and the edgeR quasi-likelihood pipeline. F1000Research 5, 1438.

9. Ritchie, M.E., Phipson, B., Wu, D., Hu, Y., Law, C.W., Shi, W., and Smyth, G.K. (2015). limma powers differential
expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Research 43(7), e47.

10. Orchestrating high-throughput genomic analysis with Bioconductor. W. Huber, V.J. Carey, R. Gentleman, ..., M.
Morgan Nature Methods, 2015:12, 115.

11. Hadley Wickham, Romain François, Lionel Henry and Kirill Müller (2021). dplyr: A Grammar of Data Manipulation.
https://dplyr.tidyverse.org, https://github.com/tidyverse/dplyr.

12. Gu, Z. (2014) circlize implements and enhances circular visualization in R. Bioinformatics.

13. Kolberg L, Raudvere U, Kuzmin I, Vilo J, Peterson H (2020). “gprofiler2- an R package for gene list functional
enrichment analysis and namespace conversion toolset g:Profiler.” _F1000Research_, *9 (ELIXIR)*(709). R package
version 0.2.1.

14. Alivernini, S., Gremese, E., McSharry, C., Tolusso, B., Ferraccioli, G., McInnes, I. B., & Kurowska-Stolarska, M. (2018). MicroRNA-155-at the Critical Interface of Innate and Adaptive Immunity in Arthritis. Frontiers in immunology, 8, 1932. https://doi.org/10.3389/fimmu.2017.01932.

15. Jacolien van Rij (2020). plotfunctions: Various Functions to Facilitate Visualization of Data and Analysis. R
package version 1.4. https://jacolienvanrij.com/tutorials.html






